### PR TITLE
IGNITE-21376 Add the catalog version in which the index was created

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21376

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java [hunk 1]
```diff
@@ -35,11 +35,15 @@ public abstract class CatalogIndexDescriptor extends CatalogObjectDescriptor {
     /** Index status. */
     private final CatalogIndexStatus status;
 
-    CatalogIndexDescriptor(int id, String name, int tableId, boolean unique, CatalogIndexStatus status) {
+    /** Catalog version in which the index was created. */
+    private final int creationVersion;
```
Comment (sashapolo) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java:39]: I would suggest to rename this parameter to `catalogCreationVersion`. We already have a mixup in "table versions" and "catalog version". Better be explicit

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerRecoveryTest.java [hunk 2]
```diff
@@ -99,6 +101,27 @@ void testRecoveryCatalogVersionTimestamps() throws Exception {
...
+    @Test
+    void testRecoveryIndexCreationCatalogVersion() throws Exception {
+        createAndStartComponents();
+
+        assertThat(catalogManager.execute(simpleTable(TABLE_NAME)), willCompleteSuccessfully());
+        assertThat(catalogManager.execute(simpleIndex(TABLE_NAME, INDEX_NAME)), willCompleteSuccessfully());
+
+        int expCreationCatalogVersion = catalogManager.latestCatalogVersion();
+
+        int indexId = catalogManager.index(INDEX_NAME, clock.now().longValue()).id();
```
Comment (sashapolo) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerRecoveryTest.java:113]: I believe there's a `clock.nowLong` method

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 3]
```diff
@@ -2266,17 +2272,65 @@ void testStartBuildingIndexEvent() {
...
+        Catalog latestCatalog = manager.catalog(manager.latestCatalogVersion());
+
+        assertThat(latestCatalog.version(), greaterThan(expCreationVersion));
+
+        assertEquals(expCreationVersion, latestCatalog.index(indexId).creationVersion());
+    }
+
+    private static Stream<Arguments> argumentsForCheckIndexCreationCatalogVersion() {
+        return Stream.of(
+                Arguments.of(true, true), // Create hash index and update index.
```
Comment (sashapolo) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2329]: What does "update index" mean?

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java (hunk 1 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogIndexDescriptor.java:39]: Ok, why not.

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerRecoveryTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerRecoveryTest.java:113]: Fix it

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2329]: Try fix it

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogSortedIndexDescriptor.java [hunk 4]
```diff
@@ -45,9 +46,10 @@ public CatalogSortedIndexDescriptor(
             String name,
             int tableId,
             boolean unique,
+            int creationVersion,
```
Comment (sashapolo) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogSortedIndexDescriptor.java:49]: You forgot to rename this one

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java [hunk 5]
```diff
@@ -2266,17 +2272,65 @@ void testStartBuildingIndexEvent() {
...
+        Catalog latestCatalog = manager.catalog(manager.latestCatalogVersion());
+
+        assertThat(latestCatalog.version(), greaterThan(expCreationVersion));
+
+        assertEquals(expCreationVersion, latestCatalog.index(indexId).creationCatalogVersion());
+    }
+
+    private static Stream<Arguments> argumentsForCheckIndexCreationCatalogVersion() {
+        return Stream.of(
+                Arguments.of(true, true), // Create hash index and update status index.
```
Comment (sashapolo) [inline modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2329]:
```suggestion
                Arguments.of(true, true), // Create hash index and update index status.
```

**Review by sashapolo — APPROVED** [review]

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogSortedIndexDescriptor.java (hunk 4 above)
Comment (tkalkirill) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/CatalogSortedIndexDescriptor.java:49]: fix it

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java (hunk 5 above)
Comment (tkalkirill) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogManagerSelfTest.java:2329]: fix it

**Pull request merged**